### PR TITLE
Local shell provisioner 

### PR DESCRIPTION
**Pull request by aaronwalker**
This provisioner makes it possible to execute a shell script from the machine running the packer build to allow for executing tools like serverspec prior to taking an image.  Currently the only way was to use the shell provisioner and run them on the remote server meaning you need to have all of the dependencies installed on the server image even if they weren't part of the image itself.

### CODE REVIEW

File: provisioner/shell-local/provisioner_test.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+  config := testConfig()
+
+  config["inline"] = []interface{}{"echo \"Hello World\""}
+  err := p.Prepare(config)
+  if err != nil {
+    t.Fatal("should not have error %s", err)
+  }
+
+  if runtime.GOOS == "windows" {
+    //this test only runs on windows
```
Comment (rasa) [inline provisioner/shell-local/provisioner_test.go]: Is the comment here correct? Shouldn't it say "this test only runs on non-windows systems"?

Comment (aaronwalker) [inline provisioner/shell-local/provisioner_test.go]: @rasa Yes....I am blame copy and paste :) I'll fix up the PR

**Comment by aaronwalker** [comment]
Seems the build failed for an unrelated reason....

**Comment by vtolstov** [comment]
Looks fine for me.

**Comment by sethvargo** [comment]
Hi @aaronwalker 

Thank you for the PR. Can you please run `go fmt` to format your code and then I can do a proper review :smile:

**Comment by johnrengelman** [comment]
Any chance this is going to get merged soon?

Also, it would be nice if `shell-local` supported the same options as the existing `shell`...that is, you could also provide a `script` or `scripts` array instead of defining everything inline.

**Comment by vtolstov** [comment]
@sethvargo please review it =), and ... merge =)

**Comment by brandonwestcott** [comment]
Would love to see this feature merged :+1:

File: provisioner/shell-local/provisioner.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+	return nil
+}
+
+func (p *Provisioner) Provision(ui packer.Ui, comm packer.Communicator) error {
+
+	scripts := make([]string, 0)
+
+	if p.config.Inline != nil {
+		tf, err := ioutil.TempFile("", "packer-shell-local")
```
Comment (sheldonh) [inline provisioner/shell-local/provisioner.go]: Consider making the temporary directory configurable. It's common to mount /tmp with the noexec option, which prevents execution of binaries in /tmp.

File: provisioner/shell-local/provisioner.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+	} else {
+		return fmt.Errorf("Inline must be defined")
+	}
+
+	for _, path := range scripts {
+		ui.Say(fmt.Sprintf("Provisioning with local shell script: %s", path))
+
+		log.Printf("Executing local script %s", path)
+
+		err := os.Chmod(path, 0770)
```
Comment (sheldonh) [inline provisioner/shell-local/provisioner.go]: This chmod is too late and too permissive. Rather restrict permissions before writing to the file, and disallow group access.

File: provisioner/shell-local/provisioner.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+		if err != nil {
+			return fmt.Errorf(
+				"Error chmodding script file to 0777 "+
```
Comment (sheldonh) [inline provisioner/shell-local/provisioner.go]: Wrong mode announced in error message.

File: provisioner/shell-local/provisioner.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+					"machine: %s", err)
+		}
+
+		out, err := exec.Command(path).Output()
+		if err != nil {
+			return fmt.Errorf("Error executeing script machine: %s", err)
```
Comment (sheldonh) [inline provisioner/shell-local/provisioner.go]: "executeing" should be "executing". "script machine" should be "shell script".

**Comment by gildas** [comment]
I am sure you guys know about https://github.com/shaunduncan/packer-provisioner-host-command that has been around for months, right?

**Comment by salizzar** [comment]
It will be great having this feature as default in Packer.

**Comment by salizzar** [comment]
Another great feature based on this is expose instance IP addresses to able SSH'ing on machine to run serverspec or a configuration management tool like Ansible..

**Comment by mitchellh** [comment]
This looks great! Instead of going with the `inline` options, I think we should just go with `command` that we execute with `sh -c` (configurable). This is pretty close. I might merge it and do that myself if you don't do it before me.

**Comment by mitchellh** [comment]
Superseded by #2283. Thanks!

**Pull request closed without merging**